crypto/src/commitment.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use alloc::vec::Vec;
+use core::fmt::Debug;
+
+use utils::{Deserializable, Serializable};
+
+use crate::Hasher;
+
+/// A vector commitment (VC) scheme.
+///
+/// This is a cryptographic primitive allowing one to commit, using a commitment string `com`, to
+/// a vector of values (v_0, ..., v_{n-1}) such that one can later reveal the value at the i-th
+/// position.
+/// This is achieved by providing the value `v_i` together with a proof `proof_i` such that anyone
+/// posessing `com` can be convinced, with high confidence, that the claim is true.
+///
+/// Vector commitment schemes usually have some batching properties in the sense that opening
+/// proofs for a number of `(i, v_i)` can be batched together into one batch opening proof in order
+/// to optimize both the proof size as well as the verification time.
+pub trait VectorCommitment<H: Hasher>: Sized {
+    /// Options defining the VC i.e., public parameters.
+    type Options: Default;
+    /// Opening proof of some value at some position index.
+    type Proof: Clone + Serializable + Deserializable;
+    /// Batch opening proof of a number of {(i, v_i)}_{i ∈ S} for an index set.
+    type MultiProof: Serializable + Deserializable;
+    /// Error returned by the scheme.
+    type Error: Debug;
+
+    /// Creates a commitment to a vector of values (v_0, ..., v_{n-1}) using the default
+    /// options.
+    fn new(items: Vec<H::Digest>) -> Result<Self, Self::Error> {
+        Self::with_options(items, Self::Options::default())
+    }
+
+    /// Creates a commitment to a vector of values (v_0, ..., v_{n-1}) given a set of
+    /// options.
+    fn with_options(items: Vec<H::Digest>, options: Self::Options) -> Result<Self, Self::Error>;
+
+    /// Returns the commitment string to the commited values.
+    fn commitment(&self) -> H::Digest;
+
+    /// Returns the length of the vector commited to for `Self::Proof`.
+    fn get_proof_domain_len(proof: &Self::Proof) -> usize;
+
+    /// Returns the length of the vector commited to for `Self::MultiProof`.
+    fn get_multiproof_domain_len(proof: &Self::MultiProof) -> usize;
```

Comment:
Not for this PR but I wonder if we should do it slightly differently: define `DomainLength` trait which could look something like this:
```rust
pub trait DomainLength {
    fn domain_length(&self) -> usize;
}
```
And then require the both `Proof` and `MultiProof` associated types implement this trait.

Reply:
Created issue